File: app/src/main/java/top/easelink/lcg/ui/main/forumnav/view/ForumNavigationFragment.java
package top.easelink.lcg.ui.main.forumnav.view;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.lifecycle.ViewModelProviders;

import org.greenrobot.eventbus.EventBus;

import javax.inject.Inject;

import top.easelink.framework.base.BaseFragment;
import top.easelink.lcg.BR;
import top.easelink.lcg.R;
import top.easelink.lcg.databinding.FragmentForumsNavigationBinding;
import top.easelink.lcg.ui.ViewModelProviderFactory;
import top.easelink.lcg.ui.main.forumnav.viewmodel.ForumNavigationViewModel;
import top.easelink.lcg.ui.main.model.ForumNavigationModel;
import top.easelink.lcg.ui.main.model.OpenForumEvent;

public class ForumNavigationFragment extends BaseFragment<FragmentForumsNavigationBinding, ForumNavigationViewModel>
        implements ForumNavigationNavigator{

    @Inject
    ViewModelProviderFactory factory;

    public static ForumNavigationFragment newInstance() {
        return new ForumNavigationFragment();
    }

    @Override
    public int getBindingVariable() {
        return BR.viewModel;
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_forums_navigation;
    }

    @Override
    public ForumNavigationViewModel getViewModel() {
        return ViewModelProviders.of(this, factory).get(ForumNavigationViewModel.class);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getViewModel().setNavigator(this);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setUp();
    }

    private void setUp() {
        getViewDataBinding().navigationGrid.setAdapter(new CustomGridViewAdapter(getBaseActivity(), R.layout.item_forums_grid));
        getViewDataBinding().navigationGrid.setOnItemClickListener(((parent, view, position, id) -> {
            ForumNavigationModel item = ((CustomGridViewAdapter) parent.getAdapter()).getItem(position);
            EventBus.getDefault().post(new OpenForumEvent(item.getTitle(), item.getUrl()));
        }));
        getViewModel().initOptions(getContext());
    }
}

File: app/src/main/java/top/easelink/lcg/ui/main/forumnav/view/ForumNavigationFragment.kt
package top.easelink.lcg.ui.main.forumnav.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import top.easelink.framework.customview.linkagerv.LinkageRecyclerView
import top.easelink.framework.topbase.ControllableFragment
import top.easelink.framework.topbase.TopFragment
import top.easelink.lcg.R
import top.easelink.lcg.ui.main.forumnav.model.ForumGroupedItem
import top.easelink.lcg.utils.getJsonStringFromAssets

class ForumNavigationFragment: TopFragment(), ControllableFragment {
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_forums_navigation_v3, container, false)
    }

    override fun isControllable(): Boolean {
        return true
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setUp()
    }

    private fun setUp() {
        view?.let {
            val json = getJsonStringFromAssets("forums.json", it.context)
            val items = try {
                Gson().fromJson<List<ForumGroupedItem>>(
                    json,
                    genericType<List<ForumGroupedItem>>()
                )
            } catch (e: Exception) {
                null
            }
            if (items != null) {
                it.findViewById<LinkageRecyclerView<ForumGroupedItem.ItemInfo>>(R.id.linkageRV)
                    .init(items, ForumsPrimaryAdapterConfig(), ForumsSecondaryAdapterConfig())

            }
        }
    }

    private inline fun <reified T> genericType() = object: TypeToken<T>() {}.type
}